lib/model/message_list.dart

```diff
@@ -524,12 +525,47 @@ class MessageListView with ChangeNotifier, _MessageSequence {
     notifyListeners();
   }
 
+  /// Update [narrow] for the result of a "with" narrow (topic permalink) fetch.
+  ///
+  /// To avoid an extra round trip, the server handles [ApiNarrowWith]
+  /// by returning results from the indicated message's current stream/topic
+  /// (if the user has access),
+  /// even if that differs from the narrow's stream/topic filters
+  /// because the message was moved.
+  ///
+  /// If such a "redirect" happened, this helper updates the stream and topic
+  /// in [narrow] to match the message's current conversation.
+  /// It also removes the "with" component from [narrow]
+  /// whether or not a redirect happened.
+  ///
+  /// See API doc:
+  ///   https://zulip.com/api/construct-narrow#message-ids
+  void _adjustNarrowForTopicPermalink(Message? someFetchedMessageOrNull) {
+    final narrow = this.narrow;
+    if (narrow is! TopicNarrow || narrow.with_ == null) return;
+
+    switch (someFetchedMessageOrNull) {
+      case null:
+        // This can't be a redirect; a redirect can't produce an empty result.
+        // (The server only redirects if the message is accessible to the user,
+        // and if it is, it'll appear in the result, making it non-empty.)
+        this.narrow = narrow.sansWith();
+      case StreamMessage():
+        this.narrow = TopicNarrow.ofMessage(someFetchedMessageOrNull);
```

Comment:
Over on the widget state there's a comment which should get updated:
```dart
  void _modelChanged() {
    if (model!.narrow != widget.narrow) {
      // A message move event occurred, where propagate mode is
      // [PropagateMode.changeAll] or [PropagateMode.changeLater].
      widget.onNarrowChanged(model!.narrow);

Reply:
Ooh thanks for the catch.